js/util.js

```diff
@@ -54,25 +54,46 @@ function setBit(bytes, n, v) {
   const mask = 128 >> Math.floor(n % 8)
   bytes[Math.floor(n / 8)] = (bytes[Math.floor(n / 8)] & ~mask) | (mask * v)
 }
-function drawOnCanvas(target, body) {
+function drawOnCanvas(target, eventPress, eventRelease=(a,b,c)=>0) {
   var mode = 0
   function drag(event) {
     if (mode == 0) { return }
+    event.preventDefault()
     const r = target.getBoundingClientRect()
-    body(
+    eventPress(
       event.clientX - r.left,
       event.clientY - r.top,
-      mode == 1
+      mode
     )
   }
-  function release(event) { mode = 0; drag(event) }
+  function release(event) {
+    if (mode == 0) { return }
+    event.preventDefault()
+    const r = target.getBoundingClientRect()
+    eventRelease(
+      event.clientX - r.left,
+      event.clientY - r.top,
+      mode
+    )
+    mode = 0
+  }
   function press  (event) { mode = event.button == 2 ? 2 : 1; drag(event) }
   function context(event) { drag(event); return false }
+  function touch(event) {
+    var touches = event.changedTouches
+    event.clientX = touches[0].pageX
+    event.clientY = touches[0].pageY
+    return event
+  }
   target.onmousemove   = drag
   target.onmouseup     = release
   target.onmouseout    = release
   target.onmousedown   = press
   target.oncontextmenu = context
+  target.addEventListener("touchmove",  e=>drag(touch(e)))
+  target.addEventListener("touchstart", e=>press(touch(e)))
+  target.addEventListener("touchend",   e=>release(touch(e)))
+  target.addEventListener("touchcancel",e=>release(touch(e)))
```

Comment:
Touch swipe to draw on canvas. Making the sprite/audio pattern data "invert" functions important for undrawing.